### PR TITLE
Include Python stack trace in crash handler dialog

### PR DESCRIPTION
**Pull request by nyalldawson**
When a crash comes from python code, we can use Python's faulthandler module to get the stack trace. We now include this in the crash handler report.

Useful for debugging, and also allows us to put the blame where it lies -- when a user script or plugin caused the crash, we clearly report that and remove the options to report to the QGIS tracker:

image

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -19,7 +19,7 @@ if (NOT IOS)
   add_subdirectory(crssync)
 endif()
 
-if(WIN32 AND NOT MINGW)
+if (NOT WITH_QT6 AND ((WIN32 AND NOT MINGW) OR (UNIX AND NOT APPLE AND NOT ANDROID)))
```
Comment (m-kuhn) [inline src/CMakeLists.txt:22]:
```suggestion
if (NOT WITH_QT6 AND ((WIN32 AND NOT MINGW) OR (UNIX AND NOT APPLE AND NOT ANDROID AND NOT IOS)))
```

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -16236,7 +16237,7 @@ void QgisApp::keyPressEvent( QKeyEvent *e )
 {
   emit keyPressed( e );
 
-#if 0 && defined(_MSC_VER) && defined(QGISDEBUG)
+#if 0 && defined(QGISDEBUG)
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp:16240]: What's this for? `#if 0` here disables this anyway?

Comment (nyalldawson) [reply src/app/qgisapp.cpp:16240]: So I can change it easily to a 1 to test 😂

Comment (m-kuhn) [reply src/app/qgisapp.cpp:16240]: So this was on purpose and now includes other platforms next to MSVC? :+1:

Comment (nyalldawson) [reply src/app/qgisapp.cpp:16240]: Yes, before the crash dialog only worked on windows

**Comment by Gustry** [comment]
Very nice idea ! Thanks a lot.

The extra-bonus would be to display a clickable link to the repository when the error is coming from the plugin :)

**Comment by nicogodet** [comment]
Maybe another issue but the trace trace order differ the current plugin stack trace error :
Your PR: most recent call first
Actual behavior : most recent call last

> Impossible de charger l'extension 'mesh_tools' provoque une erreur lors de l'appel à sa méthode classFactory() 
> 
> SyntaxError: invalid syntax 
> Traceback (most recent call last):
>   File "C:\OSGeo4W/apps/qgis/./python\qgis\utils.py", line 423, in _startPlugin
>     plugins[packageName] = package.classFactory(iface)
>   File "C:\Users/godet/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\mesh_tools\__init__.py", line 35, in classFactory
>     from .mesh_tools import MeshTools
>   File "C:\OSGeo4W/apps/qgis/./python\qgis\utils.py", line 888, in _import
>     mod = _builtin_import(name, globals, locals, fromlist, level)
>   File "C:\Users/godet/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\mesh_tools\mesh_tools.py", line 33, in 
>     from .libs.culvert_manager import CulvertManager
>   File "C:\OSGeo4W/apps/qgis/./python\qgis\utils.py", line 888, in _import
>     mod = _builtin_import(name, globals, locals, fromlist, level)
>   File "C:\Users/godet/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\mesh_tools\libs\culvert_manager.py", line 71
>     def __init__(self, parent=None)
>                                    ^
> SyntaxError: invalid syntax

**Comment by nyalldawson** [comment]
@Gustry 
>The extra-bonus would be to display a clickable link to the repository when the error is coming from the plugin :)

Yeah I was thinking this too... It's not trivial though. The problem is that the crash dialog is a seperate application and it has no access to the qgis libraries, nor any python environment available. So the whole plugin metadata scraping would need to be reimplemented in the dialog again....

**Comment by nyalldawson** [comment]
@nicogodet 
> Your PR: most recent call first
> Current behavior : most recent call last

Hmmmm, that's annoying. I'll see if I can fix this in a follow up. Unfortunately it's not as simple as reversing the lines in the log file as the faulthandler output includes other content aside from the stack

**Comment by 3nids** [comment]
really nice idea!
would you consider this for a backport to 3.22 ?
it's not really a feature and might prove to be very helpful regarding bug report process.

**Comment by nyalldawson** [comment]
@3nids maybe after a month or two testing in master first?

**Comment by 3nids** [comment]
sounds perfect

**Comment by m-kuhn** [comment]
let's create one and freeze, I like the idea to backport a lot

**Pull request merged**